sycl/source/queue.cpp

```diff
@@ -84,47 +84,66 @@ bool queue::is_host() const {
 
 void queue::throw_asynchronous() { impl->throw_asynchronous(); }
 
-event queue::memset(void *Ptr, int Value, size_t Count) {
+event queue::memset(void *Ptr, int Value,
+                    size_t Count _CODELOCPARAMDEF(&CodeLoc)) {
+  detail::tls_code_loc_t TlsCodeLocCapture(CodeLoc);
   return impl->memset(impl, Ptr, Value, Count, {});
 }
 
-event queue::memset(void *Ptr, int Value, size_t Count, event DepEvent) {
+event queue::memset(void *Ptr, int Value, size_t Count,
+                    event DepEvent _CODELOCPARAMDEF(&CodeLoc)) {
+  detail::tls_code_loc_t TlsCodeLocCapture(CodeLoc);
   return impl->memset(impl, Ptr, Value, Count, {DepEvent});
 }
 
 event queue::memset(void *Ptr, int Value, size_t Count,
-                    const std::vector<event> &DepEvents) {
+                    const std::vector<event> &DepEvents
+                        _CODELOCPARAMDEF(&CodeLoc)) {
+  detail::tls_code_loc_t TlsCodeLocCapture(CodeLoc);
   return impl->memset(impl, Ptr, Value, Count, DepEvents);
 }
 
-event queue::memcpy(void *Dest, const void *Src, size_t Count) {
+event queue::memcpy(void *Dest, const void *Src,
+                    size_t Count _CODELOCPARAMDEF(&CodeLoc)) {
+  detail::tls_code_loc_t TlsCodeLocCapture(CodeLoc);
   return impl->memcpy(impl, Dest, Src, Count, {});
 }
 
-event queue::memcpy(void *Dest, const void *Src, size_t Count, event DepEvent) {
+event queue::memcpy(void *Dest, const void *Src, size_t Count,
+                    event DepEvent _CODELOCPARAMDEF(&CodeLoc)) {
+  detail::tls_code_loc_t TlsCodeLocCapture(CodeLoc);
   return impl->memcpy(impl, Dest, Src, Count, {DepEvent});
 }
 
 event queue::memcpy(void *Dest, const void *Src, size_t Count,
-                    const std::vector<event> &DepEvents) {
+                    const std::vector<event> &DepEvents
+                        _CODELOCPARAMDEF(&CodeLoc)) {
+  detail::tls_code_loc_t TlsCodeLocCapture(CodeLoc);
   return impl->memcpy(impl, Dest, Src, Count, DepEvents);
 }
 
-event queue::mem_advise(const void *Ptr, size_t Length, pi_mem_advice Advice) {
+event queue::mem_advise(const void *Ptr, size_t Length,
+                        pi_mem_advice Advice _CODELOCPARAMDEF(&CodeLoc)) {
+  detail::tls_code_loc_t TlsCodeLocCapture(CodeLoc);
```

Comment:
Currently [mem_advise](https://github.com/intel/llvm/blob/ff4c93638654a7dd7bb6e873d9b26c986d3103cd/sycl/source/detail/queue_impl.cpp#L188) is not instrumented with XPTI. I believe that mem_advise needs to be recorded as a graph event (as in node notification) as it can result in a copy operation if the advice is set to READ_ONLY supported by a device, then something like the suggestion below should be enabled in the mem_advise implementation function referenced in the link above.

```suggestion
  #if XPTI_ENABLE_INSTRUMENTATION
  // We need a code pointer value and we'd use the object ptr; If code location
  // is available, we use the source file information along with the object
  // pointer.
  XPTIScope PrepareNotify((void *)this,
                          (uint16_t)xpti::trace_point_type_t::node_create,
                          SYCL_MEM_ALLOC_STREAM_NAME, "queue.mem_advise()");
  PrepareNotify.addMetadata([&](auto TEvent) {
    xpti::addMetadata(TEvent, "sycl_device",
                      reinterpret_cast<size_t>(
                          MDevice->is_host() ? 0 : MDevice->getHandleRef()));
    xpti::addMetadata(TEvent, "usm_memory_ptr", reinterpret_cast<size_t>(Ptr));
    xpti::addMetadata(TEvent, "memory_size", Count);
  });
  // Notify XPTI about the mem_advise submission
  PrepareNotify.notify();
  // Emit a begin/end scope for this call
  PrepareNotify.scopedNotify((uint16_t)xpti::trace_point_type_t::task_begin);
#endif
```

Reply:
@tovinkere this PR is not intended to any any extra instrumentation for XPTI. IT is needed to add code_location in API while we have a chance to do that. Any implementation gaps in terms of xpti notifications are not intended to be included here.